File: tests/fptClient.test.ts
import { jest } from '@jest/globals';
import { FptServerClient, FptRequestError } from '../src/fptClient.js';

function makeMockAxios(handlers: { get?: jest.Mock; post?: jest.Mock }) {
  const requestInterceptors: any[] = [];
  const responseInterceptors: any[] = [];
  return {
    get: handlers.get ?? jest.fn(),
    post: handlers.post ?? jest.fn(),
    interceptors: {
      request: { use: (fn: any) => requestInterceptors.push(fn) },
      response: { use: (onFulfilled: any, onRejected: any) => responseInterceptors.push({ onFulfilled, onRejected }) },
    },
    _requestInterceptors: requestInterceptors,
    _responseInterceptors: responseInterceptors,
  } as any;
}

describe('FptServerClient', () => {
  test('caches /actions GET responses but not other paths', async () => {
    const get = jest.fn()
      .mockResolvedValueOnce({ data: { actions: ['a'] }, headers: {} })
      .mockResolvedValueOnce({ data: { jobs: [] }, headers: {} })
      .mockResolvedValueOnce({ data: { jobs: [] }, headers: {} });
    const client = new FptServerClient(makeMockAxios({ get }));

    await client.get('/actions');
    await client.get('/actions'); // should hit cache, not call axios again
    await client.get('/jobs');
    await client.get('/jobs'); // uncached path, should call axios again

    expect(get).toHaveBeenCalledTimes(3);
  });

  test('post clears the cache so a subsequent GET refetches', async () => {
    const get = jest.fn()
      .mockResolvedValueOnce({ data: { actions: ['a'] }, headers: {} })
      .mockResolvedValueOnce({ data: { actions: ['b'] }, headers: {} });
    const post = jest.fn().mockResolvedValueOnce({ data: { ok: true }, headers: {} });
    const client = new FptServerClient(makeMockAxios({ get, post }));

    await client.get('/actions');
    await client.post('/jobs/x/cancel');
    const second = await client.get<any>('/actions');

    expect(get).toHaveBeenCalledTimes(2);
    expect(second.actions).toEqual(['b']);
  });

  test('wraps an error response into FptRequestError with the API code/message', async () => {
    const axiosMock = makeMockAxios({});
    new FptServerClient(axiosMock);
    const [{ onRejected }] = axiosMock._responseInterceptors;

    const apiError = {
      response: {
        status: 409,
        data: { error: { code: 'job.already_finished', message: 'Build đã kết thúc' } },
      },
      message: 'Request failed with status code 409',
    };

    await expect(onRejected(apiError)).rejects.toMatchObject({
      status: 409,
      code: 'job.already_finished',
      message: 'Build đã kết thúc',
    });
    await expect(onRejected(apiError)).rejects.toBeInstanceOf(FptRequestError);
  });
});

File: test/fpt_client_test.dart
import 'dart:convert';

import 'package:fpt_server_mcp/src/fpt_client.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

void main() {
  group('FptClient', () {
    test('caches /actions GET responses but not other paths', () async {
      var getCount = 0;
      final client = FptClient(
        baseUrl: 'https://example.test',
        apiKey: '',
        client: MockClient((request) async {
          if (request.url.path == '/actions') {
            getCount++;
            return http.Response(jsonEncode({'actions': []}), 200);
          }
          getCount++;
          return http.Response(jsonEncode({'jobs': []}), 200);
        }),
      );

      await client.getJson('/actions');
      await client.getJson('/actions'); // should hit cache
      await client.getJson('/jobs');
      await client.getJson('/jobs'); // uncached path, calls again

      expect(getCount, 3);
    });

    test('post clears the cache so a subsequent GET refetches', () async {
      var actionsBody = 'a';
      var getCount = 0;
      final client = FptClient(
        baseUrl: 'https://example.test',
        apiKey: '',
        client: MockClient((request) async {
          if (request.method == 'POST') {
            actionsBody = 'b';
            return http.Response(jsonEncode({'ok': true}), 200);
          }
          getCount++;
          return http.Response(jsonEncode({'actions': actionsBody}), 200);
        }),
      );

      await client.getJson('/actions');
      await client.postJson('/jobs/x/cancel');
      final second = await client.getJson('/actions');

      expect(getCount, 2);
      expect(second['actions'], 'b');
    });

    test(
        'wraps an error response into FptRequestError with the API code/message',
        () async {
      final client = FptClient(
        baseUrl: 'https://example.test',
        apiKey: '',
        client: MockClient(
          (request) async => http.Response.bytes(
            utf8.encode(
              jsonEncode({
                'error': {
                  'code': 'job.already_finished',
                  'message': 'Build đã kết thúc',
                },
              }),
            ),
            409,
            headers: {'content-type': 'application/json; charset=utf-8'},
          ),
        ),
      );

      await expectLater(
        client.getJson('/jobs/x'),
        throwsA(
          isA<FptRequestError>()
              .having((e) => e.status, 'status', 409)
              .having((e) => e.code, 'code', 'job.already_finished')
              .having((e) => e.message, 'message', 'Build đã kết thúc'),
        ),
      );
    });

    test('sends the API key as X-API-Key when configured', () async {
      String? seenHeader;
      final client = FptClient(
        baseUrl: 'https://example.test',
        apiKey: 'secret-key',
        client: MockClient((request) async {
          seenHeader = request.headers['X-API-Key'];
          return http.Response(jsonEncode({'ok': true}), 200);
        }),
      );

      await client.getJson('/health');
      expect(seenHeader, 'secret-key');
    });
  });
}
